lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1621,8 +1621,9 @@ def _change_datatype(
         if isinstance(item, HistoryDatasetAssociation):
             self.hda_manager.ensure_can_change_datatype(item)
             self.hda_manager.ensure_can_set_metadata(item)
-            item.dataset.state = item.dataset.states.SETTING_METADATA
-            trans.sa_session.flush()
+            if not item.has_deferred_data:
+                item.dataset.state = item.dataset.states.SETTING_METADATA
+                trans.sa_session.flush()
             change_datatype.delay(dataset_id=item.id, datatype=params.datatype)
```

Comment:
```suggestion
                trans.sa_session.flush()
                change_datatype.delay(dataset_id=item.id, datatype=params.datatype)
```
My comment below only makes sense with that change ...

Reply:
sorry, I should take more time to read this all of this.